### PR TITLE
CBMs for NPCs 2: Active CBMs with AI control

### PR DESCRIPTION
**Pull request by mlangsdorf**
#### Summary
```SUMMARY: Features "CBMs for NPCs 2: Active CBMs with AI control"```

#### Purpose of change
Some CBMs can be installed on NPCs, but they have to be passive effects because the NPC AI doesn't know how to turn on CBMs or use them.

Teach the NPC AI to use active CBMs, and make a wider range of CBMs available to NPCs.

#### Describe the solution
Create a bunch of utility functions to control when NPCs turn on and off CBMs.  Use those functions to  allow NPCs to use CBMs.

Add dialogue options to control NPC active CBM usage somewhat.

#### Describe alternatives you've considered
I strongly considered creating a map of installed CBMs to bionics index values within npc::, but C++ map semantics are so bad that there's really no benefit to doing that.

I'd like to figure out how to get NPCs to use power armor interfaces and internal UPSes, but it evaded me in this pass.

I'm not sure if the offensive versus defensive/utility power reserve makes sense.  I am undecided on the recharge mechanics: currently, NPCs will recharge up a limit, which I think makes the most sense.  The limit can be set between 10% and 90% of total bionic power.

#### Additional context
It works, but needs more playtesting and refactoring for submission.

I still need to update the NPC tutorial, but I want to finalize the reserve and recharge mechanics first.

### CODE REVIEW

File: src/npcmove.cpp [hunk 1]
```diff
@@ -1304,16 +1396,210 @@ const item_location npc::find_usable_ammo( const item &weap ) const
...
+        }
+    }
+}
+
+void npc::deactivate_combat_cbms()
+{
+    for( const bionic_id &cbm_id : defense_cbms ) {
+        deactivate_bionic_by_id( cbm_id );
+    }
+    deactivate_bionic_by_id( bio_hydraulics );
```
Comment (Night-Pryanik) [inline src/npcmove.cpp]: It seems that hydraulics are only deactivated, but ain't activated in `activate_combat_cbms()` above.

File: src/npcmove.cpp [hunk 2]
```diff
@@ -1304,16 +1396,210 @@ const item_location npc::find_usable_ammo( const item &weap ) const
...
+        return true;
+    }
+
+add_msg("mjll %s is trying to recharge", name);
+    use_bionic_by_id( bio_torsionratchet );
+    use_bionic_by_id( bio_metabolics );
+
+    if( use_bionic_by_id( bio_furnace ) ) {
+        const std::function<bool( const item & )> furnace_filter = []( const item & it ) {
+            return it.typeId() == itype_id( "withered" ) || it.typeId() == itype_id( "files" ) ||
```
Comment (Night-Pryanik) [inline src/npcmove.cpp]:
```suggestion
            return it.typeId() == itype_id( "withered" ) || it.typeId() == itype_id( "file" ) ||
```

File: src/npcmove.cpp (hunk 1 above)
Comment (mlangsdorf) [reply src/npcmove.cpp]: Hydraulics are activated when you're actually making reach or melee attacks, so as to limit unnecessary power drain when moving or if you're using ranged weapons.

**Comment by Mcrone617** [comment]
Hey @mlangsdorf I'm working with @DracoGriffin and we've been trying to get public participation up on the subreddit (and in general) by having people testing the PR's. Now your coding and work is already amazing for cataclysm development, and the work you've been doing on the NPC's especially impressive. You noted that this PR needs some playtesting, how about releasing a downloadable build and having the community try it out?

This public testing serves a dual purpose also, as it would give people a reason to see and use all the new NPC interactions you have been working on. I'm thinking some sort of challenge to create a cyborg NPC ally, possibly via debug, and have people mess around with that for a couple of in-game days, to test the active bionics, and to just get people more familiar with the new NPC functions.

All this would require on your end is a release of a playable build, preferably via GitHub, but any distribution network really works. Of course if you're interested, feel free to post the PR on the subreddit yourself, but if you're too busy, then just release a download link, and I'll post it and think of a appropriate challenge for it. Let me know if you're interested, i'm on the discord also, so feel free to message me there if you have questions. Hope to hear back from you!

**Comment by mlangsdorf** [comment]
I would be happy to have public playtesting, but I have no idea how to create a release for general distribution.  Is there a document somewhere for that?

**Comment by Mcrone617** [comment]
> I would be happy to have public playtesting, but I have no idea how to create a release for general distribution. Is there a document somewhere for that?

I've mostly been helping people out individually, but firestorm was the first to figure it out, so hopefully his advice helps you find out what the problem is.
"Every fork has "release" tab. There you can find "Draft new release" And you can actually upload built files there manually."
"My link is just release from my fork on github. It was built by me personally."
"To author. You can upload you build directly to github.

While drafting new release you can simply upload you files in archive

https://i.imgur.com/PFVcwdc.png "

after this, under your release tab there should be a easy to distribute downloadable link. you do have to make a build with your changes, but it's apparently not that difficult. Let me know if you run into any problems, and I'll try to help out. It is a shame that there is no formal documentation, I might try to make a formal guide in a couple of weeks once I have more time to make this system more clear.

**Pull request merged**